jzaia_files/main.cpp

```diff
@@ -0,0 +1,225 @@
+                                         control_values,
+                                         {num_qubits-1},
+                                         false);
+
+    // Apply H to all non-ancilla qubits
+    for (size_t i = 0; i < num_qubits-1; ++i) {
+        GateImplementationsLM::applyHadamard(sv.getData(),
+                                             sv.getNumQubits(),
+                                             {i},
+                                             false);
+    }
+}
+
+/**
+ * @brief The first testing oracle for Grover's
+ *
+ * A 6-qubit test oracle for Grover's algorithm. Applies a pauli-X to the
+ * rightmost qubit if the leftmost 5 qubits are in the state |11010>
+ *
+ * @param sv The statevector to apply the oracle to. Must be 6 qubits
+ * @param control_vals The state the oracle should select
+ */
+void oracle(StateVectorLQubitManaged<double> &sv,
+            const std::vector<bool> &control_vals) {
+    // Sanity check statevector
+    assert(sv.getNumQubits() == control_vals.size()+1);
+
+    // Define controls to be used for applying the X gate
+    std::vector<size_t> controls(control_vals.size());
+    std::iota(controls.begin(), controls.end(), 0);
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {control_vals.size()},
+                                         false);
+}
+
+/**
+ * @brief Overall function for running Grover's algorithm on a chosen oracle
+ *
+ * Run Grover's algorithm from start to finish. Prepares a statevector, and
+ * repeats state selection and amplitude-amplification for sqrt(N) iterations
+ * (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+ * the oracle always picks precisly 1 state (rather than an arbitrary number).
+ *
+ * @param num_qubits The number of qubits in the circuit the oracle acts
+ *        on (includes the ancilla)
+ * @param expected The state which the oracle should apply phase to
+ */
+void run_experiment(const size_t num_qubits,
+                    const std::vector<bool> &expected) {
+    // Create the statevector for this circuit
+    // NOTE: qubits are numbered left-to-right order
+    StateVectorLQubitManaged<double> sv(num_qubits);
+
+    // Set up the circuit by placing into uniform superposition
+    groversSetup(sv);
+
+    // Apply (approx) sqrt(N) repetitions of state selection and amp-amp
+    const size_t nreps = static_cast<size_t>(sqrt(1llu << (num_qubits-1)));
+    for (size_t reps = nreps; reps > 0; --reps) {
+        // Apply the oracle to apply a phase of -1 to desired state
+        oracle(sv, expected);
+        // Perform amp-amp by reflecting over |+++...+>
+        groversMirror(sv);
+    }
```

Comment:
What do you think about looping forward instead of backward?

Reply:
That would also be completely reasonable. Originally `nreps` was inline and not its own variable (i.e `reps = <the code that currently initializes nreps>`), and I formulated the loop this way so that it wouldn't recompute `nreps` on every loop iteration (though the compiler would likely optimize this away if it was able anyway). In it's current state, forward vs backward iteration shouldn't make any difference, even including bad compiler optimization. Changed to forward iteration for readability in 0066b3b.